packages/broker/src/plugins/operator/ContractFacade.ts

```diff
@@ -0,0 +1,371 @@
+                    `
+            }
+        }
+        const parseItems = (response: any) => {
+            if (!response.operator) {
+                logger.error('Unable to find operator in The Graph', { operatorContractAddress: this.operatorContract.address })
+                return []
+            }
+            return response.operator.stakes
+        }
+        this.theGraphClient.updateRequiredBlockNumber(requiredBlockNumber)
+        return this.theGraphClient.queryEntities<{ id: string, sponsorship: { id: string, stream: { id: string } } }>(createQuery, parseItems)
+    }
+
+    addReviewRequestListener(listener: ReviewRequestListener, abortSignal: AbortSignal): void {
+        addManagedEventListener<any, any>(
+            this.operatorContract as any,
+            'ReviewRequest',
+            (sponsorship: string, targetOperator: string, metadataAsString?: string) => {
+                let partition: number
+                try {
+                    partition = parsePartitionFromReviewRequestMetadata(metadataAsString)
+                } catch (err) {
+                    if (err instanceof ParseError) {
+                        logger.warn(`Skip review request (${err.reasonText})`, {
+                            address: this.operatorContract.address,
+                            sponsorship,
+                            targetOperator,
+                        })
+                    } else {
+                        logger.warn('Encountered unexpected error', { err })
+                    }
+                    return
+                }
+                logger.debug('Receive review request', {
+                    address: this.operatorContract.address,
+                    sponsorship,
+                    targetOperator,
+                    partition
+                })
+                listener(toEthereumAddress(sponsorship), toEthereumAddress(targetOperator), partition)
+            },
+            abortSignal
+        )
+    }
+
+    async getStreamId(sponsorshipAddress: string): Promise<StreamID> {
+        const sponsorship = new Contract(sponsorshipAddress, sponsorshipABI, this.config.signer) as unknown as Sponsorship
+        return toStreamID(await sponsorship.streamId())
+    }
+
+    async voteOnFlag(sponsorship: string, targetOperator: string, kick: boolean): Promise<void> {
+        const voteData = kick ? VOTE_KICK : VOTE_NO_KICK
+        await (await this.operatorContract.voteOnFlag(sponsorship, targetOperator, voteData)).wait()
+    }
+
+    addOperatorContractStakeEventListener(eventName: 'Staked' | 'Unstaked', listener: (sponsorship: string) => unknown): void {
+        this.operatorContract.on(eventName, listener)
+    }
+
+    removeOperatorContractStakeEventListener(eventName: 'Staked' | 'Unstaked', listener: (sponsorship: string) => unknown): void {
+        this.operatorContract.off(eventName, listener)
+    }
+
+    getProvider(): Provider {
+        return this.config.signer.provider!
+    }
```

Comment:
`!` means trouble, sooner or later. It's better to face the facts.

```ts
getProvider(): Provider {
    const { provider } = this.config.signer

    if (!provider) {
        throw new Error('I splode here so others dont have to')
    }

    return provider
}

```

Reply:
I would actually check for this in the constructor of the class. That way we fail even faster and basically make it clear that "we only accept signers that have an associated provider".

Reply:
Created a PR about this: https://github.com/streamr-dev/network/pull/1954. The client always provides a signer with `Provider`, and after this PR our type definitions also ensures it.